### ISSUE TITLE
Regression in 2.1.9: excess duplicate dataloader fetches

### ISSUE THREAD

**Issue by myronmarston** [rmosolgo/graphql-ruby#4747]
**Describe the bug**

When attempting to upgrade to 2.1.9, I noticed that `fetch` on my dataloader gets called some extra times with arguments I already resolved.

**Versions**

`graphql` version: 2.1.9
`rails` (or other framework): N/A
`ruby` version: 3.2.2

**GraphQL schema**

```graphql
type ManufacturerConnection {
  nodes: [Manufacturer]
}

type Manufacturer {
  manufactured_parts: PartConnection
}

type PartConnection {
  nodes: [Part]
}

union Part = ElectricalPart | MechanicalPart

type ElectricalPart {
  voltage: Int
  components(filter: ComponentFilter): ComponentConnection
}

type MechanicalPart {
  weight: Int
  components(filter: ComponentFilter): ComponentConnection
}

type ComponentConnection {
  nodes: [Component]
}

type Component {
  widget: Widget
}

type Widget {
  name: String
}

input ComponentFilter {
  name: StringFilter
}

input StringFilter {
  equal_to_any_of: [String]
}

type Query {
  manufacturers: ManufacturerConnection
}
```

**GraphQL query**

```graphql
query {
  manufacturers {
    nodes {
      manufactured_parts {
        nodes {
          ... on MechanicalPart {
            components(filter: {name: {equal_to_any_of: ["c1", "c2", "c3"]}}) {
              nodes {
                widget {
                  name
                }
              }
            }
          }
          ... on ElectricalPart {
            components(filter: {name: {equal_to_any_of: ["c1", "c2", "c3"]}}) {
              nodes {
                widget {
                  name
                }
              }
            }
          }
        }
      }
    }
  }
}
```

**Steps to reproduce**

Put this into a ruby script:

```ruby
require "bundler/inline"

gemfile do
  source "https://rubygems.org"
  gem "graphql", ENV.fetch("GRAPHQL_VERSION")
end

require "graphql"
require "graphql/dataloader/source"

SCHEMA_STRING = <<~EOS
  type ManufacturerConnection {
    nodes: [Manufacturer]
  }

  type Manufacturer {
    manufactured_parts: PartConnection
  }

  type PartConnection {
    nodes: [Part]
  }

  union Part = ElectricalPart | MechanicalPart

  type ElectricalPart {
    voltage: Int
    components(filter: ComponentFilter): ComponentConnection
  }

  type MechanicalPart {
    weight: Int
    components(filter: ComponentFilter): ComponentConnection
  }

  type ComponentConnection {
    nodes: [Component]
  }

  type Component {
    widget: Widget
  }

  type Widget {
    name: String
  }

  input ComponentFilter {
    name: StringFilter
  }

  input StringFilter {
    equal_to_any_of: [String]
  }

  type Query {
    manufacturers: ManufacturerConnection
  }
EOS

class Application
  def initialize
    @schema = GraphQL::Schema.from_definition(SCHEMA_STRING, default_resolve: self, using: {GraphQL::Dataloader => {}})
  end

  def call(parent_type, field, object, args, context)
    case field.name
    when "manufacturers", "manufactured_parts", "components", "widget"
      context.dataloader.with(FieldSource).load(FieldAndObject.new(field, object))
    when "nodes"
      object
    when "name"
      object.first.fetch("name")
    end
  end

  def resolve_type(supertype, object, context)
    possible_types = @schema.possible_types(supertype)
    possible_types[object.fetch("id") % possible_types.size]
  end

  def execute_query(query_string)
    @schema.execute(query_string)
  end

  class FieldSource < GraphQL::Dataloader::Source
    def fetch(fields)
      puts "Bulk loading data for #{fields.size} fields: #{fields.map(&:description).join(", ")}"

      fields.map do
        [
          {"id" => 1, "name" => "a"},
          {"id" => 2, "name" => "b"},
          {"id" => 3, "name" => "c"},
          {"id" => 4, "name" => "d"},
        ]
      end
    end
  end

  FieldAndObject = ::Data.define(:field, :object) do
    def description
      "#{field.name}(for #{object&.fetch("id").inspect})"
    end
  end
end

components_field =
  if ENV["OMIT_FILTER"]
    "components"
  else
    'components(filter: {name: {equal_to_any_of: ["c1", "c2", "c3"]}})'
  end

app = Application.new

query = <<~QUERY
  query {
    manufacturers {
      nodes {
        manufactured_parts {
          nodes {
            ... on MechanicalPart {
              #{components_field} {
                nodes {
                  widget {
                    name
                  }
                }
              }
            }
            ... on ElectricalPart {
              #{components_field} {
                nodes {
                  widget {
                    name
                  }
                }
              }
            }
          }
        }
      }
    }
  }
QUERY

response = app.execute_query(query).to_h

puts
puts "Response:"
puts response
```

**Expected behavior**

On GraphQL 2.1.8, here's how it behaves:

```
$ GRAPHQL_VERSION=2.1.8 ruby graphql_gem_bug.rb
Bulk loading data for 1 fields: manufacturers(for nil)
Bulk loading data for 4 fields: manufactured_parts(for 1), manufactured_parts(for 2), manufactured_parts(for 3), manufactured_parts(for 4)
Bulk loading data for 4 fields: components(for 1), components(for 2), components(for 3), components(for 4)
Bulk loading data for 4 fields: widget(for 1), widget(for 2), widget(for 3), widget(for 4)

Response:
{"data"=>{"manufacturers"=>{"nodes"=>[{"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}]}}}
```

I expect to see the same output when running against 2.1.9.

**Actual behavior**

Here's the output on 2.1.9:

```
$ GRAPHQL_VERSION=2.1.9 ruby graphql_gem_bug.rb
Bulk loading data for 1 fields: manufacturers(for nil)
Bulk loading data for 4 fields: manufactured_parts(for 1), manufactured_parts(for 2), manufactured_parts(for 3), manufactured_parts(for 4)
Bulk loading data for 4 fields: components(for 1), components(for 2), components(for 3), components(for 4)
Bulk loading data for 4 fields: widget(for 1), widget(for 2), widget(for 3), widget(for 4)
Bulk loading data for 3 fields: components(for 1), components(for 2), components(for 3)

Response:
{"data"=>{"manufacturers"=>{"nodes"=>[{"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}]}}}
```

Notice the extra `Bulk loading data for 3 fields: components(for 1), components(for 2), components(for 3)` line in the output. That shows that `fetch` on my dataloader is being called an extra time, with arguments that were already resolved before.

**Additional context**

The presence of arguments in the query seems to make a difference, even when my resolver logic ignores them. If I omit the arguments, then it behaves like 2.1.8 did:

```
$ GRAPHQL_VERSION=2.1.9 OMIT_FILTER=true ruby graphql_gem_bug.rb
Bulk loading data for 1 fields: manufacturers(for nil)
Bulk loading data for 4 fields: manufactured_parts(for 1), manufactured_parts(for 2), manufactured_parts(for 3), manufactured_parts(for 4)
Bulk loading data for 4 fields: components(for 1), components(for 2), components(for 3), components(for 4)
Bulk loading data for 4 fields: widget(for 1), widget(for 2), widget(for 3), widget(for 4)

Response:
{"data"=>{"manufacturers"=>{"nodes"=>[{"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}, {"manufactured_parts"=>{"nodes"=>[{"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}, {"components"=>{"nodes"=>[{"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}, {"widget"=>{"name"=>"a"}}]}}]}}]}}}
```

It's possible that this has the same root cause as #4745 and a fix for that issue may fix this issue but it could be a different bug so I wanted to open up a separate issue.

**Comment by rmosolgo** [comment]
Thanks for the really detailed report. I have a hunch that you're right, that the algorithm changes I made in #4743 had unexpected (and untested 😖 ) changes... I'm going to work up tests in the test suite for this case and the other one you reported, then step back recent Dataloader changes until I can understand _why_. Thanks so much for your help in tracking this down!

**Comment by rmosolgo** [comment]
I just released GraphQL-Ruby 2.1.10 which reverts the default Fiber control approach to use `resume/Fiber.yield`. Could you try that version and let me know how it goes for you?

**Comment by myronmarston** [comment]
> I just released GraphQL-Ruby 2.1.10 which reverts the default Fiber control approach to use `resume/Fiber.yield`. Could you try that version and let me know how it goes for you?

It fixed it for me. Thanks!
